File: day-two.py
# My original version

print('My original version')

with open('day-two.txt') as f:
    num_seq = f.read().split(',')

num_seq = [tuple(map(int, num.split('-'))) for num in num_seq]
part_one_invalid_seq = []
part_two_invalid_seq = []

def part_one_is_invalid(n):
    s = str(n)
    mid = len(s) // 2
    return len(s) % 2 == 0 and s[:mid] == s[mid:]

def part_two_is_invalid(n):
    s = str(n)

    for divided in range(2, len(s) + 1):
        if len(s) % divided == 0:
            part = s[: len(s) // divided]
            if part * divided == s:
                return True
    return False

for start_num, end_num in num_seq:
    for num in range(start_num, end_num + 1):
        if part_one_is_invalid(num):
            part_one_invalid_seq.append(num)
        if part_two_is_invalid(num):
            part_two_invalid_seq.append(num)

print(sum(part_one_invalid_seq))
print(sum(part_two_invalid_seq))

# Claude improved version

print('Claude improved version')

def parse_ranges(filename: str) -> list[tuple[int, int]]:
    with open(filename) as f:
        ranges = f.read().split(',')
    return [tuple(map(int, r.split('-'))) for r in ranges]

def is_repeated_twice(n: int) -> bool:
    s = str(n)
    length = len(s)
    
    if length % 2 != 0:
        return False
    
    mid = length // 2
    return s[:mid] == s[mid:]

def is_repeated(n: int) -> bool:
    s = str(n)
    length = len(s)
    
    # Try all possible pattern lengths (from 1 to length/2)
    for pattern_length in range(1, length // 2 + 1):
        if length % pattern_length == 0:
            pattern = s[:pattern_length]
            repeated_times = length // pattern_length
            if pattern * repeated_times == s:
                return True

    return False

def find_invalid_ids(ranges: list[tuple[int, int]], validator) -> list[int]:
    invalid_ids = []
    
    for start, end in ranges:
        for num in range(start, end + 1):
            if validator(num):
                invalid_ids.append(num)
    
    return invalid_ids

def main() -> None:
    ranges = parse_ranges('day-two.txt')
    
    part_one_invalid = find_invalid_ids(ranges, is_repeated_twice)
    part_two_invalid = find_invalid_ids(ranges, is_repeated)
    
    print(sum(part_one_invalid))
    print(sum(part_two_invalid))

if __name__ == '__main__':
    main()

File: day-two.exs
defmodule DayTwo do
  @moduledoc """
  Finds and sums invalid IDs based on repeated digit patterns.
  """

  def parse_ranges(filename) do
    filename
    |> File.read!()
    |> String.trim()
    |> String.split(",")
    |> Enum.map(&parse_range/1)
  end

  defp parse_range(range) do
    [start, stop] = String.split(range, "-")
    {String.to_integer(start), String.to_integer(stop)}
  end

  def repeated_twice?(n) do
    str = Integer.to_string(n)
    len = String.length(str)
    half = div(len, 2)

    rem(len, 2) == 0 and String.slice(str, 0, half) == String.slice(str, half, half)
  end

  def repeated?(n) do
    str = Integer.to_string(n)
    len = String.length(str)

    len > 1 and
      Enum.any?(1..div(len, 2), &has_repeating_pattern?(str, len, &1))
  end

  defp has_repeating_pattern?(str, len, pattern_len) do
    if rem(len, pattern_len) == 0 do
      pattern = String.slice(str, 0, pattern_len)
      repetitions = div(len, pattern_len)
      String.duplicate(pattern, repetitions) == str
    else
      false
    end
  end

  def find_invalid_ids(ranges, validator) do
    ranges
    |> Enum.flat_map(fn {start, stop} ->
      start..stop |> Enum.filter(validator)
    end)
  end

  def main do
    ranges = parse_ranges("day-two.txt")

    part_one_sum = ranges |> find_invalid_ids(&repeated_twice?/1) |> Enum.sum()
    part_two_sum = ranges |> find_invalid_ids(&repeated?/1) |> Enum.sum()

    IO.puts(part_one_sum)
    IO.puts(part_two_sum)
  end
end

DayTwo.main()
